src/Xamarin.Android.Build.Tasks/Utilities/XAAssemblyResolver.cs

```diff
@@ -0,0 +1,333 @@
+	{
+		this.log = log;
+		this.loadDebugSymbols = loadDebugSymbols;
+		this.readerParameters = loadReaderParameters ?? new ReaderParameters();
+
+		cache = new Dictionary<string, CacheEntry> (StringComparer.OrdinalIgnoreCase);
+	}
+
+	public AssemblyDefinition Resolve (string fullName, ReaderParameters? parameters = null)
+	{
+		return Resolve (AssemblyNameReference.Parse (fullName), parameters);
+	}
+
+	public AssemblyDefinition Resolve (AssemblyNameReference name)
+	{
+		return Resolve (name, null);
+	}
+
+	public AssemblyDefinition Resolve (AssemblyNameReference name, ReaderParameters? parameters)
+	{
+		return Resolve (AndroidTargetArch.None, name, parameters);
+	}
+
+	public AssemblyDefinition Resolve (AndroidTargetArch arch, AssemblyNameReference name, ReaderParameters? parameters = null)
+	{
+		string shortName = name.Name;
+		if (cache.TryGetValue (shortName, out CacheEntry? entry)) {
+			return SelectAssembly (arch, name.FullName, entry, loading: false);
+		}
+
+		if (arch == AndroidTargetArch.None) {
+			return FindAndLoadFromDirectories (arch, FrameworkSearchDirectories, name, parameters);
+		}
+
+		if (!AbiSearchDirectories.TryGetValue (arch, out ICollection<string>? directories) || directories == null) {
+			throw CreateLoadException (name);
+		}
+
+		return FindAndLoadFromDirectories (arch, directories, name, parameters);
+	}
+
+	AssemblyDefinition FindAndLoadFromDirectories (AndroidTargetArch arch, ICollection<string> directories, AssemblyNameReference name, ReaderParameters? parameters)
+	{
+		string? assemblyFile;
+		AssemblyDefinition? candidate = null;
+		foreach (string dir in directories) {
+			if ((assemblyFile = SearchDirectory (name.Name, dir)) != null) {
+				AssemblyDefinition? loaded = Load (arch, assemblyFile);
+				if (Array.Equals (loaded?.Name.MetadataToken, name.MetadataToken)) {
+					return loaded;
+				}
+				candidate = candidate ?? loaded;
+			}
+		}
+
+		if (candidate != null)
+			return candidate;
+
+		throw CreateLoadException (name);
+	}
+
+	static FileNotFoundException CreateLoadException (AssemblyNameReference name)
+	{
+		return new FileNotFoundException ($"Could not load assembly '{name}'.");
+	}
+
+	static string? SearchDirectory (string name, string directory)
+	{
+		if (Path.IsPathRooted (name) && File.Exists (name)) {
+			return name;
+		}
+
+		var file = Path.Combine (directory, $"{name}.dll");
+		if (File.Exists (file)) {
+			return file;
+		}
+
+		return null;
+	}
+
+	public virtual AssemblyDefinition? Load (AndroidTargetArch arch, string filePath, bool forceLoad = false)
+	{
+		string name = Path.GetFileNameWithoutExtension (filePath);
+		AssemblyDefinition? assembly;
+		if (!forceLoad && cache.TryGetValue (name, out CacheEntry? entry)) {
+			assembly = SelectAssembly (arch, name, entry, loading: true);
+			if (assembly != null) {
+				return assembly;
+			}
+		}
+
+		try {
+			assembly = ReadAssembly (filePath);
```

Comment:
Related to [`ReaderParameters? parametesr`](https://github.com/xamarin/xamarin-android/pull/8164/files#r1261618106), shouldn't `ReadAssembly()` take the `parameters` value as well?